File: frontend/web/user/src/services/table.service.ts
import axios from "axios";
import type {
  AddScheduleRequest,
  AddScheduleResponse,
  CurrentScheduleResponse,
  CurrentTermSubjectsResponse,
  DeleteScheduleResponse,
  ScheduleItem,
  ScheduleDetailResponse,
  UpdateScheduleRequest,
  UpdateScheduleResponse,
} from "@/interfaces/table.interface";
import { authenticatedApiClient } from "./api.client";

class TableService {
  private readonly apiClient = authenticatedApiClient;

  async getCurrentSchedule(): Promise<CurrentScheduleResponse | null> {
    try {
      const response = await this.apiClient.get<CurrentScheduleResponse>(
        "/user/schedule"
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response?.status === 404) {
        return null;
      }
      throw this.toError(error, "Unable to load schedule");
    }
  }

  async getScheduleDetail(scheduleTimeId: number): Promise<ScheduleItem> {
    try {
      const response = await this.apiClient.get<ScheduleDetailResponse>(
        `/user/schedule/detail/${scheduleTimeId}`
      );
      return response.data.data;
    } catch (error) {
      throw this.toError(error, "Unable to load schedule details");
    }
  }

  async getCurrentTermSubjects(): Promise<CurrentTermSubjectsResponse> {
    try {
      const response = await this.apiClient.get<CurrentTermSubjectsResponse>(
        "/user/schedule/subjects"
      );
      return response.data;
    } catch (error) {
      throw this.toError(error, "Unable to load current term subjects");
    }
  }

  async addSchedule(data: AddScheduleRequest): Promise<AddScheduleResponse> {
    try {
      const response = await this.apiClient.post<AddScheduleResponse>(
        "/user/schedule/add-time",
        data
      );
      return response.data;
    } catch (error) {
      throw this.toError(error, "Unable to add schedule block");
    }
  }

  async updateSchedule(
    scheduleTimeId: number,
    data: UpdateScheduleRequest
  ): Promise<UpdateScheduleResponse> {
    try {
      const response = await this.apiClient.put<UpdateScheduleResponse>(
        `/user/schedule/edit/${scheduleTimeId}`,
        data
      );
      return response.data;
    } catch (error) {
      throw this.toError(error, "Unable to update schedule");
    }
  }

  async deleteSchedule(scheduleTimeId: number): Promise<DeleteScheduleResponse> {
    try {
      const response = await this.apiClient.delete<DeleteScheduleResponse>(
        `/user/schedule/${scheduleTimeId}`
      );
      return response.data;
    } catch (error) {
      throw this.toError(error, "Unable to delete schedule");
    }
  }

  private toError(error: unknown, fallbackMessage: string): Error {
    if (axios.isAxiosError(error)) {
      const message = error.response?.data?.message || error.message;
      return new Error(message || fallbackMessage);
    }
    return error instanceof Error ? error : new Error(fallbackMessage);
  }
}

export const tableService = new TableService();
export default tableService;

File: frontend/mobile/lib/services/table.service.dart
import 'package:dio/dio.dart';

import '../interfaces/table.interface.dart';
import 'api.service.dart';

class TableException implements Exception {
  final String message;

  const TableException(this.message);

  @override
  String toString() => message;
}

abstract class TableRepository {
  Future<CurrentSchedule?> getCurrentSchedule();
  Future<ScheduleItem> getScheduleDetail(int scheduleTimeId);
  Future<List<ScheduleSubject>> getCurrentTermSubjects();
  Future<void> addSchedule(AddScheduleInput input);
  Future<void> updateSchedule(int scheduleTimeId, UpdateScheduleInput input);
  Future<void> deleteSchedule(int scheduleTimeId);
}

class TableService implements TableRepository {
  final ApiService _apiService;

  TableService({ApiService? apiService})
    : _apiService = apiService ?? ApiService();

  @override
  Future<CurrentSchedule?> getCurrentSchedule() async {
    try {
      final response = await _apiService.get('/user/schedule');
      return CurrentSchedule.fromJson(
        Map<String, dynamic>.from(response.data as Map),
      );
    } on DioException catch (error) {
      if (error.response?.statusCode == 404) return null;
      throw _exception(error, 'ไม่สามารถโหลดตารางเวลาได้');
    } on TypeError {
      throw const TableException('ข้อมูลตารางเวลาจากระบบไม่ถูกต้อง');
    }
  }

  @override
  Future<ScheduleItem> getScheduleDetail(int scheduleTimeId) async {
    try {
      final response = await _apiService.get(
        '/user/schedule/detail/$scheduleTimeId',
      );
      final body = Map<String, dynamic>.from(response.data as Map);
      return ScheduleItem.fromJson(
        Map<String, dynamic>.from(body['data'] as Map),
      );
    } on DioException catch (error) {
      throw _exception(error, 'ไม่สามารถโหลดรายละเอียดตารางเวลาได้');
    } on TypeError {
      throw const TableException('ข้อมูลรายละเอียดตารางเวลาไม่ถูกต้อง');
    }
  }

  @override
  Future<List<ScheduleSubject>> getCurrentTermSubjects() async {
    try {
      final response = await _apiService.get('/user/schedule/subjects');
      final body = Map<String, dynamic>.from(response.data as Map);
      final data = body['data'];
      if (data is! List) return const [];
      return data
          .whereType<Map>()
          .map(
            (item) => ScheduleSubject.fromJson(Map<String, dynamic>.from(item)),
          )
          .toList();
    } on DioException catch (error) {
      throw _exception(error, 'ไม่สามารถโหลดรายวิชาได้');
    }
  }

  @override
  Future<void> addSchedule(AddScheduleInput input) async {
    try {
      await _apiService.post('/user/schedule/add-time', data: input.toJson());
    } on DioException catch (error) {
      throw _exception(error, 'ไม่สามารถเพิ่มบล็อกเวลาได้');
    }
  }

  @override
  Future<void> updateSchedule(
    int scheduleTimeId,
    UpdateScheduleInput input,
  ) async {
    try {
      await _apiService.put(
        '/user/schedule/edit/$scheduleTimeId',
        data: input.toJson(),
      );
    } on DioException catch (error) {
      throw _exception(error, 'ไม่สามารถแก้ไขตารางเวลาได้');
    }
  }

  @override
  Future<void> deleteSchedule(int scheduleTimeId) async {
    try {
      await _apiService.delete('/user/schedule/$scheduleTimeId');
    } on DioException catch (error) {
      throw _exception(error, 'ไม่สามารถลบบล็อกเวลาได้');
    }
  }

  TableException _exception(DioException error, String fallback) {
    final data = error.response?.data;
    if (data is Map && data['message'] is String) {
      return TableException(data['message'] as String);
    }
    return TableException(fallback);
  }
}
